README.md
# HTTPInterceptor

[![Platform](https://img.shields.io/badge/Platform-iOS-lightgrey.svg?colorA=28a745&colorB=4E4E4E)](https://img.shields.io/badge/Platform-iOS-lightgrey.svg?colorA=28a745&colorB=4E4E4E)
[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/HTTPInterceptor.svg?style=flat&label=CocoaPods&colorA=28a745&&colorB=4E4E4E)](https://cocoapods.org/pods/HTTPInterceptor)
[![Swift support](https://img.shields.io/badge/Swift-4.0%20%7C%204.1%20%7C%204.2%20%7C%205.0-lightgrey.svg?colorA=28a745&colorB=4E4E4E)](#swift-versions-support)
[![Build Status](https://travis-ci.org/ladeiko/HTTPInterceptor.svg?branch=master)](https://travis-ci.org/ladeiko/HTTPInterceptor)

## Introduction

Helps to setup interceptor for HTTP(s) requests, modify them, return custom response.

## Changelog

See [CHANGELOG](CHANGELOG.md)

## Installation

### Cocoapods
> This is the recommended way of installing this package.

* Add the following line to your Podfile

``` ruby
pod 'HTTPInterceptor'
```
* Run the following command to fetch and build your dependencies

``` bash
pod install
```

### Manually
If you prefer to install this package manually, just follow these steps:

* Make sure your project is a git repository. If it isn't, just run this command from your project root folder:

``` bash
git init
```

* Add HTTPInterceptor as a git submodule by running the following command.

``` bash
git submodules add https://github.com/ladeiko/HTTPInterceptor.git
```
* Add files from *'submodules/HTTPInterceptor/Sources'* folder to your project.

## Usage

See usage examples in [HTTPInterceptorDemoTests.swift](Demo/HTTPInterceptorDemoTests/HTTPInterceptorDemoTests.swift)

## LICENSE
This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
# CHANGELOG

## 1.0.2
### Added
* Support for mapping of folders by subpath of URL, e.g: _http://google.com/sub/a_ -> _http://google.com/sub_ 'a'

## 1.0.0
* Initial version



Sources/HTTPInterceptor.h
//
//  HTTPInterceptor.h
//
//  Created by Siarhei Ladzeika on 16 Apr 19.
//  Copyright © 2019 Siarhei Ladzeika. All rights reserved.
//

#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

typedef void (^NSURLProtocolInterceptorRequestPreprocessor)(NSMutableURLRequest* request);

typedef void (^NSURLProtocolInterceptorFailureDataCompletion)(NSError* error);
typedef void (^NSURLProtocolInterceptorSuccessDataCompletion)(NSInteger httpStatusCode, id __nullable data, NSString* __nullable mimeType, NSString* __nullable encoding, NSDictionary* __nullable headerFields);
typedef void (^NSURLProtocolInterceptorRequestFilter)(NSURLRequest* request, NSURLProtocolInterceptorSuccessDataCompletion successCompletion, NSURLProtocolInterceptorFailureDataCompletion __nullablefailureCompletion);

typedef NSString* NSURLProtocolInterceptorHandlerKey;

@interface HTTPInterceptor : NSObject

+ (NSURLProtocolInterceptorHandlerKey)addPreprocessor:(NSURLProtocolInterceptorRequestPreprocessor)preprocessor;
+ (void)removePreprocessorForKey:(NSURLProtocolInterceptorHandlerKey)key;
+ (NSURLProtocolInterceptorHandlerKey)addInterceptor:(NSURLProtocolInterceptorRequestFilter)filter;
+ (void)removeInterceptorForKey:(NSURLProtocolInterceptorHandlerKey)key;

+ (NSURLProtocolInterceptorHandlerKey)mapUrl:(NSURL*)url toLocalPathUrl:(NSURL*)localUrl;

@end

NS_ASSUME_NONNULL_END


Demo/HTTPInterceptorDemo/AppDelegate.swift
//
//  AppDelegate.swift
//  NSURLProtocolDemo
//
//  Created by Siarhei Ladzeika on 4/15/19.
//  Copyright © 2019 Siarhei Ladzeika. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Demo/HTTPInterceptorDemoTests/HTTPInterceptorDemoTests.swift
//
//  HTTPInterceptorDemoTests.swift
//  HTTPInterceptorDemoTests
//
//  Created by Siarhei Ladzeika on 4/16/19.
//  Copyright © 2019 Siarhei Ladzeika. All rights reserved.
//

import XCTest
import HTTPInterceptor

class HTTPInterceptorDemoTests: XCTestCase {

    func testRequestPreprocessor() {
        
        let expectedFieldName = "X-" + UUID().uuidString
        let expectedFieldValue = UUID().uuidString
        
        let key = HTTPInterceptor.addPreprocessor { (request) in
            if let url = request.url, url.absoluteString == "https://postman-echo.com/get" {
                request.setValue(expectedFieldValue, forHTTPHeaderField: expectedFieldName)
            }
        }
        
        struct PostmanResponse: Decodable {
            var args: [String: String]
            var headers: [String: String]
            var url: String
        }
        
        let decorder = JSONDecoder()
        
        let data1 = try! Data(contentsOf: URL(string: "https://postman-echo.com/get")!)
        let response1 = try! decorder.decode(PostmanResponse.self, from: data1)
        
        XCTAssert(response1.headers[expectedFieldName.lowercased()] == expectedFieldValue)
        
        HTTPInterceptor.removePreprocessor(forKey: key)
        
        let data2 = try! Data(contentsOf: URL(string: "https://postman-echo.com/get")!)
        let response2 = try! decorder.decode(PostmanResponse.self, from: data2)
        
        XCTAssert(response2.headers[expectedFieldName.lowercased()] == nil)
    }

    func testRequestInterceptor() {
        
        let expectedResponse = UUID().uuidString
        var called = 0

        let key = HTTPInterceptor.add { (request, succeeded, failed) in
            if let url = request.url, url.absoluteString == "https://google.com" {
                called += 1
                succeeded(200, expectedResponse.data(using: .utf8), "text/html", "utf-8", nil)
            }
        }
        
        let interceptedActualResponse = try! String(contentsOf: URL(string: "https://google.com")!, encoding: .utf8)
        XCTAssert(expectedResponse == interceptedActualResponse)
        XCTAssert(called == 1)
        
        HTTPInterceptor.remove(forKey: key)
        
        let realActualResponse = try! String(contentsOf: URL(string: "https://google.com")!, encoding: .utf8)
        XCTAssert(expectedResponse != realActualResponse)
        XCTAssert(called == 1)
    }
    
    func testFileMapping() {
        
        let localURL = URL(fileURLWithPath: NSTemporaryDirectory().appending(UUID().uuidString))
        
        self.addTeardownBlock {
            try? FileManager.default.removeItem(at: localURL)
        }
        
        try! FileManager.default.createDirectory(at: localURL, withIntermediateDirectories: true, attributes: nil)
        
        let content = UUID().uuidString
        try! content.write(to: localURL.appendingPathComponent("content.txt"), atomically: true, encoding: .utf8)
        
        let key = HTTPInterceptor.mapUrl(URL(string: "http://hello.com")!, toLocalPathUrl: localURL.appendingPathComponent("content.txt"))
        
        let actual1 = try! String(contentsOf: URL(string: "http://hello.com")!)
        XCTAssert(actual1 == content)
        
        let actual2 = try! String(contentsOf: URL(string: "http://hello.com/")!)
        XCTAssert(actual2 == content)
        
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com:80")!))
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com/a")!))
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com/content2.txt")!))

        HTTPInterceptor.remove(forKey: key)
    }
    
    func testFolderMapping() {
        
        let localURL = URL(fileURLWithPath: NSTemporaryDirectory().appending(UUID().uuidString))
        
        self.addTeardownBlock {
            try? FileManager.default.removeItem(at: localURL)
        }
        
        try! FileManager.default.createDirectory(at: localURL, withIntermediateDirectories: true, attributes: nil)
        
        let content1 = UUID().uuidString
        try! content1.write(to: localURL.appendingPathComponent("content1.txt"), atomically: true, encoding: .utf8)
        
        let content2 = UUID().uuidString
        try! content2.write(to: localURL.appendingPathComponent("content2.txt"), atomically: true, encoding: .utf8)
        
        let key = HTTPInterceptor.mapUrl(URL(string: "http://hello.com")!, toLocalPathUrl: localURL)
        
        let actual1 = try! String(contentsOf: URL(string: "http://hello.com/content1.txt")!)
        XCTAssert(actual1 == content1)
        
        let actual2 = try! String(contentsOf: URL(string: "http://hello.com/content2.txt")!)
        XCTAssert(actual2 == content2)
        
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com/content3.txt")!))
        
        HTTPInterceptor.remove(forKey: key)
    }
    
    func testFolderWithSubpathMapping() {
        
        let localURL = URL(fileURLWithPath: NSTemporaryDirectory().appending(UUID().uuidString))
        
        self.addTeardownBlock {
            try? FileManager.default.removeItem(at: localURL)
        }
        
        try! FileManager.default.createDirectory(at: localURL, withIntermediateDirectories: true, attributes: nil)
        
        let content1 = UUID().uuidString
        try! content1.write(to: localURL.appendingPathComponent("content1.txt"), atomically: true, encoding: .utf8)
        
        let content2 = UUID().uuidString
        try! content2.write(to: localURL.appendingPathComponent("content2.txt"), atomically: true, encoding: .utf8)
        
        let key = HTTPInterceptor.mapUrl(URL(string: "http://hello.com/sub")!, toLocalPathUrl: localURL)
        
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com")!))
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com/")!))
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com/sub")!))
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com/sub/")!))
        
        let actual1 = try! String(contentsOf: URL(string: "http://hello.com/sub/content1.txt")!)
        XCTAssert(actual1 == content1)
        
        let actual2 = try! String(contentsOf: URL(string: "http://hello.com/sub/content2.txt")!)
        XCTAssert(actual2 == content2)
        
        XCTAssertThrowsError(try String(contentsOf: URL(string: "http://hello.com/sub/content3.txt")!))
        
        HTTPInterceptor.remove(forKey: key)
    }

}


Sources/HTTPInterceptor.m
//
//  HTTPInterceptor.m
//
//  Created by Siarhei Ladzeika on 16 Apr 19.
//  Copyright © 2017 Siarhei Ladzeika. All rights reserved.
//

#import "HTTPInterceptor.h"
#import <CoreServices/CoreServices.h>

static BOOL urlsAreEqual(NSURL* a, NSURL* b) {
    return [[a URLByAppendingPathComponent:@""] isEqual:[b URLByAppendingPathComponent:@""]];
}

static NSString * const NSURLProtocolInterceptorHandledKey = @"NSURLProtocolInterceptorHandledKey";
static NSMutableArray<NSDictionary*>* gRequestFilters = nil;
static NSMutableArray<NSDictionary*>* g_preprocessor = nil;

@interface NSURLProtocolInterceptor : NSURLProtocol<NSURLConnectionDelegate> {}
@property (nonatomic, strong) NSURLConnection *connection;
@end

@implementation HTTPInterceptor

+ (NSURLProtocolInterceptorHandlerKey)addPreprocessor:(__nonnull NSURLProtocolInterceptorRequestPreprocessor)preprocessor {
    NSString* const key = [@"PRE-" stringByAppendingString:[[NSUUID UUID] UUIDString]];
    @synchronized(self) {
        if (!g_preprocessor) {
            g_preprocessor = [NSMutableArray new];
        }
        [g_preprocessor addObject:@{ @"key": key, @"block": [preprocessor copy] }];
    }
    return key;
}

+ (void)removePreprocessorForKey:(__nonnull NSURLProtocolInterceptorHandlerKey)key {
    @synchronized(self) {
        [g_preprocessor filterUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
            return ![[evaluatedObject objectForKey:@"key"] isEqualToString:key];
        }]];
    }
}

+ (NSString*)addInterceptor:(NSURLProtocolInterceptorRequestFilter)filter {
    NSString* const key = [@"REQ-" stringByAppendingString:[[NSUUID UUID] UUIDString]];
    @synchronized(self) {
        if (!gRequestFilters) {
            gRequestFilters = [NSMutableArray new];
        }
        [gRequestFilters addObject:@{ @"key": key, @"block": [filter copy] }];
    }
    return key;
}


+ (void)removeInterceptorForKey:(NSString*)key {
    @synchronized(self) {
        [gRequestFilters filterUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(id  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
            return ![[evaluatedObject objectForKey:@"key"] isEqualToString:key];
        }]];
    }
}

+ (NSURLProtocolInterceptorHandlerKey)mapUrl:(NSURL*)url toLocalPathUrl:(NSURL*)localUrl {
    NSString* const target = url.absoluteString;
    return [self addInterceptor:^(NSURLRequest * _Nonnull request, NSURLProtocolInterceptorSuccessDataCompletion  _Nonnull successCompletion, NSURLProtocolInterceptorFailureDataCompletion  _Nonnull __nullablefailureCompletion) {
        
        if (![request.URL.absoluteString hasPrefix:target]){
            return;
        }
        
        const BOOL match = urlsAreEqual(url, request.URL);
        
        void(^const f404)(void) = ^ {
            successCompletion(404, nil, nil, nil, nil);
        };
        
        NSString* localFilePath = nil;
        
        BOOL isDirectory = NO;

        if ([[NSFileManager defaultManager] fileExistsAtPath:localUrl.path isDirectory:&isDirectory] && !isDirectory) {
            if (match) {
                localFilePath = localUrl.path;
            }
            else {
                f404();
                return;
            }
        }
        else {
            
            NSURLComponents* const targetComponents = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
            NSString* targetPath = [targetComponents path];
            
            if ([targetPath hasSuffix:@"/"]) {
                targetPath = [targetPath substringToIndex:[targetPath length] - 1];
            }

            NSURLComponents* const components = [NSURLComponents componentsWithURL:request.URL resolvingAgainstBaseURL:NO];
            NSString* path = [components path];
            
            if ([path hasSuffix:@"/"]) {
                path = [path substringToIndex:[path length] - 1];
            }
            
            path = [path substringFromIndex:[targetPath length]];
            
            if ([path hasPrefix:@"/"]) {
                path = [path substringFromIndex:1];
            }
            
            localFilePath = [path length] ? [[localUrl URLByAppendingPathComponent:path] path] : [localUrl path];
            
            if (![[NSFileManager defaultManager] fileExistsAtPath:localFilePath isDirectory:&isDirectory]) {
                f404();
                return;
            }
            
            if (isDirectory) {
                localFilePath = [localFilePath stringByAppendingPathComponent:@"index.html"];
                if (![[NSFileManager defaultManager] fileExistsAtPath:localFilePath isDirectory:&isDirectory] || isDirectory) {
                    f404();
                    return;
                }
            }
        }
        
        NSData* const data = [NSData dataWithContentsOfFile:localFilePath];
        if (!data) {
            f404();
            return;
        }
        
        NSString* const fileExtension = [localFilePath pathExtension];
        
        NSString* const UTI = (__bridge_transfer NSString *)UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, (__bridge CFStringRef)fileExtension, NULL);
        NSString* const contentType = (__bridge_transfer NSString *)UTTypeCopyPreferredTagWithClass((__bridge CFStringRef)UTI, kUTTagClassMIMEType);
        NSString* encoding = nil;
        
        if ([@[@"text"] containsObject:[[contentType componentsSeparatedByString:@"/"] firstObject]]) {
            encoding = @"utf-8";
        }
        
        successCompletion(200, [NSURL fileURLWithPath:localFilePath], contentType, encoding, nil);
    }];
}

@end

@implementation NSURLProtocolInterceptor

+ (void)load {
    [NSURLProtocol registerClass:NSURLProtocolInterceptor.class];
}

+ (BOOL)canInitWithRequest:(NSURLRequest *)request {
    
    NSString* const lc = [request.URL.absoluteString lowercaseString];
    
    if (![lc hasPrefix:@"http://"] && ![lc hasPrefix:@"https://"]) {
        return NO;
    }
    
    if ([NSURLProtocol propertyForKey:NSURLProtocolInterceptorHandledKey inRequest:request]) {
        return NO;
    }
    
    return YES;
}

+ (NSURLRequest *) canonicalRequestForRequest:(NSURLRequest *)request {
    return request;
}

- (void) startLoading {
    
    NSMutableURLRequest *newRequest = [self.request mutableCopy];
    
    [NSURLProtocol setProperty:@YES forKey:NSURLProtocolInterceptorHandledKey inRequest:newRequest];
    
    NSArray<NSDictionary*>* preprocessors;
    
    @synchronized([HTTPInterceptor class]) {
        preprocessors = [g_preprocessor copy];
    }
    
    for (NSDictionary* info in preprocessors) {
        const NSURLProtocolInterceptorRequestPreprocessor f = (NSURLProtocolInterceptorRequestPreprocessor)info[@"block"];
        f(newRequest);
    }
    
    NSArray<NSDictionary*>* filters;
    
    @synchronized([HTTPInterceptor class]) {
        filters = [gRequestFilters copy];
    }
    
    if ([filters count] > 0 && [newRequest.URL.absoluteString hasPrefix:@"http"]) {
        
        __block BOOL completed = NO;
        
        const NSURLProtocolInterceptorSuccessDataCompletion successCompletion = ^(NSInteger httpStatusCode, id data, NSString* mimeType, NSString* encoding, NSDictionary* extraHeaderFields){
            assert(completed == NO);
            assert(httpStatusCode > 0 && httpStatusCode < 600);
            completed = YES;
            
            unsigned long long contentLength = 0;
            
            if ([data isKindOfClass:[NSData class]]) {
                contentLength = [data length];
            }
            else if ([data isKindOfClass:[NSURL class]]) {
                NSURL* const url = (NSURL*)data;
                contentLength = [[[NSFileManager defaultManager] attributesOfItemAtPath:url.path error:NULL] fileSize];
            }
            
            NSMutableDictionary* const headerFields = extraHeaderFields ? [extraHeaderFields mutableCopy] : [NSMutableDictionary new];
            NSMutableArray* contentType = [NSMutableArray new];
            
            [headerFields setObject: [@(contentLength) stringValue] forKey:@"Content-Length"];
            
            if (mimeType) {
                [contentType addObject:mimeType];
            }
            
            if (encoding) {
                if ([encoding isEqualToString:@"utf8"]) {
                    encoding = @"utf-8";
                }
                [contentType addObject:[NSString stringWithFormat:@"charset=%@", encoding]];
            }
            
            if ([contentType count] > 0) {
                [headerFields setObject:[contentType componentsJoinedByString:@"; "] forKey:@"Content-Type"];
            }
            
            NSHTTPURLResponse* const response = [[NSHTTPURLResponse alloc] initWithURL:self.request.URL
                                                                            statusCode:httpStatusCode
                                                                           HTTPVersion:@"HTTP/1.1"
                                                                          headerFields:[headerFields copy]];
            [self.client URLProtocol:self didReceiveResponse:response cacheStoragePolicy:NSURLCacheStorageNotAllowed];
            
            if ([data isKindOfClass:[NSData class]]) {
                [self.client URLProtocol:self didLoadData:data];
            }
            else if ([data isKindOfClass:[NSURL class]]) {
                @autoreleasepool {
                    NSURL* const url = data;
                    NSFileHandle* handle = [NSFileHandle fileHandleForReadingAtPath:url.path];
                    static const NSInteger limit = 1024 * 1024;
                    while (true) {
                        @autoreleasepool {
                        NSData* data = [handle readDataOfLength:limit];
                            if ([data length] > 0) {
                                [self.client URLProtocol:self didLoadData:data];
                            }
                            else {
                                break;
                            }
                        }
                    }
                }
            }
            
            [self.client URLProtocolDidFinishLoading:self];
        };
        
        const NSURLProtocolInterceptorFailureDataCompletion failureCompletion = ^(NSError* error) {
            assert(completed == NO);
            completed = YES;
            [self.client URLProtocol:self didFailWithError:error];
        };
        
        for (NSDictionary* info in filters) {
            const NSURLProtocolInterceptorRequestFilter f = (NSURLProtocolInterceptorRequestFilter)info[@"block"];
            f(self.request, successCompletion, failureCompletion);
            if (completed){
                return;
            }
        }
    }
    
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
    self.connection = [NSURLConnection connectionWithRequest:newRequest delegate:self];
    #pragma clang diagnostic pop
}

- (void) stopLoading {
    [self.connection cancel];
}

#pragma mark - NSURLConnectionDelegate

- (void) connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    [self.client URLProtocol:self didReceiveResponse:response cacheStoragePolicy:NSURLCacheStorageNotAllowed];
}

- (void) connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    [self.client URLProtocol:self didLoadData:data];
}

- (void) connectionDidFinishLoading:(NSURLConnection *)connection {
    [self.client URLProtocolDidFinishLoading:self];
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
    [self.client URLProtocol:self didFailWithError:error];
}

@end


